### PR TITLE
feat: notification center server phase (composed read-model, cursor, routes)

### LINKED ISSUE #1353: Notification center Phase 1: server (composed read-model, cursor, routes)
Phase 1 of the notification center. Spec: [`docs/design/notification-center.md`](https://github.com/ms2sato/agent-console/blob/main/docs/design/notification-center.md) (merged in #1352 — normative).

**Server only. No client work in this phase** (that is Phase 2), so this can run alongside client-side work in other Issues.

## Scope

- `NotificationItem` shared type + **valibot schema** (`strictObject`) per spec §8.
- The two v1 composers per spec §4's admission rule: **artifact created** and **worktree deletion reached `completed` / `stalled`**. Composed from the existing domain rows at request time — this phase writes no notification rows (§3).
- `user_notification_cursor (user_id PK, last_seen_at)` table + migration (§3, §5).
- `GET /api/notifications` → `{ items, lastSeenAt }`, newest first, capped at 50, no pagination.
- `PUT /api/notifications/seen { lastSeenAt }` — **monotonic**: the server rejects moving the cursor backwards.
- **Personal feed** (§8, owner-confirmed): artifacts by `user_id`, deletion jobs by payload `requestUsername`.

## The invariants this phase must not break

- **N1** — the response is built from domain rows read at request time. Nothing in this phase may cache or materialize a notification payload; a WebSocket broadcast is a hint to refetch, never a source of list content. This is what makes [#1330](https://github.com/ms2sato/agent-console/issues/1330)'s failure mode structurally unrepresentable.
- **N2** — items carry no state of their own beyond the render summary.

## Verification floor

- `pre-pr-completeness.md` **Q10** applies: type + schema in the same PR, plus a wire-boundary integration test in `packages/integration/` (a TS-only field is silently stripped by valibot).
- Boundary values stated up front per `design-principles.md`: empty feed (no artifacts, no deletions), cursor absent (never opened the bell), cursor newer than every item, an item exactly at the cursor timestamp, and the cap boundary (>50 items).
- Migration test alongside the existing `migration-vNN.test.ts` series; sibling tests per `test-trigger.md`.

## Not in this phase

The bell / panel / badge UI (Phase 2); any additional source (§4 rule 1 admits them only once they have durable records); the obligations lane (§6); a global team feed (§8).

---

## Acceptance Criteria

Spec: `docs/design/notification-center.md` (normative). Invariants N1/N2 binding throughout.

> **Spec tension found at AC time**: §8's wire is `{ items, lastSeenAt }` while its client bullet says "badge = server-computed unread count". Since `items` is capped at 50, a client-side count **lies once unread exceeds the cap**. Ruling: the GET response carries `unreadCount`, computed server-side **pre-cap**, and spec §8 gains a one-line amendment (correction-trail style) in this same PR.

### Rulings (the four open calls)

**R1 — Composer placement: `services/notification-service.ts` (flat-first), NOT the repository layer.** Repositories are per-table single-store accessors; a cross-store composer there would blur that boundary. The "read-model, not a store" property is enforced STRUCTURALLY by the service's dependency surface:

- Constructor takes READ interfaces only: `artifactRepository`, the job queue's read side (`getJobs`), and `NotificationCursorRepository` (the one writer it may use, for the cursor only).
- **Its deps interface has NO broadcast/WS/event member** — N1's "a broadcast is never a source of list content" becomes unrepresentable in the type, not merely forbidden in prose.
- Per-source composers live as an internal registry: `const SOURCES: ReadonlyArray<(identity: FeedIdentity, deps) => Promise<NotificationItem[]>>` — adding a source is one entry at one reviewable location (the §4 admission rule's enforcement point).

**R2 — Cursor monotonicity: structural in the repository's SQL, thin in the route.** The cursor repository's single write is a conditional upsert (`ON CONFLICT DO UPDATE ... WHERE excluded.last_seen_at > current`), so a backward PUT is a **no-op by construction**, not a checked convention. Semantics: idempotent no-op (NOT 409) — the response returns the CURRENT stored cursor so a stale tab converges instead of erroring. The route layer adds only: valibot format validation (400) and **future-timestamp rejection** (400 when beyond server now + a small tolerance) — a future cursor would silently swallow all subsequent notifications forever, the one non-monotonic footgun the SQL condition cannot catch. Tests: backward no-op proven by read-back; two concurrent PUTs converge to the max.

**R3 — Identity asymmetry: accepted for v1, contained at ONE seam.** Define `FeedIdentity { userId, username }`, resolved once from `authUser` at the route and passed to every source composer; each composer declares which key it filters by (artifacts → `userId`; deletion jobs → `username`). This contains the #1339 smell — no per-source ad-hoc identity derivation can accrete, because the seam is the only identity input. Do NOT build a username→userId reverse-resolution layer now (over-engineering for one source). Record the residual in the PR body: `requestUsername` is an attribution **proxy** (it exists for elevation, not attribution); the durable fix — future job payloads carrying the initiating `users.id` — triggers when a THIRD username-keyed source appears or any username-remapping feature lands. Edge pinned: a job payload with `requestUsername: null` matches nobody (the route always passes a username today; null = legacy/edge rows, excluded, documented).

**R4 — Boundary values (additions to the Issue's list):**

1. **Tie at the cursor**: unread = `occurredAt > lastSeenAt` STRICTLY (an item exactly at the cursor is seen — required for "PUT with the newest item's timestamp" to zero the badge).
2. **Tie between items**: deterministic secondary sort `(occurredAt desc, kind, id)` — two sources emitting the same timestamp must render in a stable order.
3. **Mixed-source timestamp normalization**: artifacts carry ISO strings, jobs carry epoch ms — normalize to one comparison basis for sort/unread math; the wire stays ISO.
4. **Corrupt job payload**: a `worktree:delete` row whose payload fails to narrow → skip-with-warn, never 500 (the #1347 `deriveFromJob` + DataIntegrityError-containment convention: one bad row must not kill the feed).
5. **Non-terminal jobs** (`pending` / `processing`) are absent from the feed (the kind is `worktree-deletion-finished`).
6. **Cap applied after merge+sort** (fetch up to 50 per source, merge, sort, cap 50) — note in the PR body that a chatty source can crowd the other out beyond the cap; accepted v1 cut.
7. `unreadCount` computed pre-cap (see the wire ruling above); boundary test: 51+ unread → `unreadCount` exceeds `items.length`.
8. Invalid `lastSeenAt` format → 400.

### S — Implementation

- **S1**: shared `NotificationItem` type + strictObject schema per spec §8 **plus `unreadCount` on the GET response envelope** (spec §8 amended one line, correction-trail style, same PR).
- **S2**: `user_notification_cursor` migration (next free version — v31 if nothing lands first; verify at implementation time, #1351 carries no migration) + `schema.ts` + `migration-vNN.test.ts` per the series' convention.
- **S3**: `NotificationCursorRepository` (interface + sqlite impl; the conditional-upsert single writer) per R2.
- **S4**: `notification-service.ts` per R1, with the two v1 composers.
- **S5**: `routes/notifications.ts` (GET + PUT) mounted inside `/api` (auth inherited; route test asserts unauthenticated 401), app-context wiring.

### T — Tests

Every R4 boundary as a named test; N1 containment (the service's deps type has no broadcast member — a compile-level fact, asserted via a type-level test or a deps-shape unit assertion); **Q10 wire-boundary integration test** in `packages/integration/` (real route → JSON → schema parse → items + `unreadCount` + `lastSeenAt` all survive); route auth test; migration test.

### File manifest (parallel dispatch with #1351 — collision map)

This PR touches: `packages/shared/src/types/notification.ts` (new) / `schemas/notification.ts` (new) / **`schemas/index.ts` + `shared/src/index.ts` export lines (the ONLY files #1351 may also touch — trivial add-line merges)** / server: `database/connection.ts`, `database/schema.ts`, new repository + service + route files, `routes/api.ts` (mount line), `app-context.ts` (wiring) / `packages/integration/` (new test file).

It must NOT touch: `mcp-server.ts`, `session-manager.ts`, `schemas/embedded-agent.ts`, or anything under `packages/client/` — that is #1351's territory.

### Non-goals

Everything the Issue lists, plus: no WS broadcast additions in this phase (the invalidation-hint wiring is Phase 2's client work consuming EXISTING broadcasts; if Phase 2 finds that no existing broadcast fires for artifact creation, that gap comes back as its own decision — do not pre-build one here).

### PR DESCRIPTION
**Pull request by ms2sato**
## Summary

Phase 1 (server-only) of the notification center: a bell + list feature that lets a human find out what happened while they were not watching, composed at request time from existing domain rows — never from a stored notification table.

- `NotificationItem` shared type + `strictObject` valibot schema (`types/notification-item.ts` / `schemas/notification-item.ts` — new files, distinct from the pre-existing unrelated outbound-Slack `notification.ts` pair, see naming-collision note below).
- Two v1 composers (§4 admission rule): artifact created, worktree deletion reached `completed`/`stalled`.
- `user_notification_cursor (user_id PK, last_seen_at)` table via migration v31.
- `GET /api/notifications` → `{ items, lastSeenAt, unreadCount }`, newest first, capped at 50, no pagination.
- `PUT /api/notifications/seen { lastSeenAt }` — monotonic (SQL-level conditional upsert; a backward/equal move is a no-op, not an error).

Spec: `docs/design/notification-center.md` (normative, merged in #1352). Rulings R1–R4 and the `unreadCount` wire amendment come from the Issue's AC, authored by the Architect.

## Invariants (structural, not conventional)

- **N1**: `NotificationServiceDeps` (in `services/notification-service.ts`) has exactly three members — `artifactRepository`, `jobs` (the job queue's read side), `cursorRepository` — and **no broadcast/WS/event member at all**. "A broadcast is never a source of list content" is therefore unrepresentable in the type. Asserted by a deps-shape unit test (`Object.keys(deps)` exact-match).
- **N2**: notification items carry no state of their own beyond the render summary — nothing is persisted per item.

## R1–R4 rulings, as implemented

- **R1**: composer placement is `services/notification-service.ts` (flat-first), not the repository layer. Per-source composers live in an internal `SOURCES` registry — adding a source is one entry at one reviewable location.
- **R2**: cursor monotonicity is structural in `SqliteNotificationCursorRepository.advance`'s SQL (`ON CONFLICT DO UPDATE ... WHERE last_seen_at < excluded value`) — a backward PUT is a no-op by construction, not a checked convention. The route returns the *current* stored cursor (idempotent, not 409). The route layer adds valibot format validation (400) and future-timestamp rejection (400, 5s tolerance) — the one footgun the SQL condition can't catch.
- **R3**: `FeedIdentity { userId, username }` resolved once at the route from `authUser`, passed to every composer; artifacts filter by `userId`, deletion jobs by payload `requestUsername`. `requestUsername: null` matches nobody (documented edge case, legacy/edge rows).

  **Residual note (R3, as required by the AC):** `requestUsername` is an attribution **proxy** — it exists for privilege elevation, not identity attribution. The durable fix — future job payloads carrying the initiating `users.id` — triggers when a THIRD username-keyed source appears, or any username-remapping feature lands. (Mirrored as a code comment at the filtering line in `composeWorktreeDeletionFinished`.)
- **R4**: all 8 named boundaries — see Test plan below.

## Spec amendment (correction-trail, not silent divergence)

`docs/design/notification-center.md` §8 amended: `GET /api/notifications`'s envelope now documents `unreadCount`, computed **pre-cap** so the badge stays accurate once unread exceeds the 50-item cap (items alone can't express this since it's truncated).

## Naming collision found and resolved (architect-approved mid-implementation)

The AC's file manifest named `packages/shared/src/types/notification.ts` / `schemas/notification.ts` as new files for this feature's wire type/schema. Both files already existed on `main`, unrelated to this feature — they hold the OUTBOUND (Slack) notification system's types from PR #201 (`OutboundTriggerEventType`, `RepositorySlackIntegration`, etc.). Consulted the Architect directly; resolution: the new wire type/schema live in new files `types/notification-item.ts` / `schemas/notification-item.ts` instead, and both the old and new files gained a one-line JSDoc cross-reference so a future reader doesn't grab the wrong one. The existing `notification.ts` pair's content is otherwise untouched.

## Collision map with #1351 (parallel PR)

This PR touches only `packages/shared/src/index.ts` and `packages/shared/src/schemas/index.ts` in common with #1351's territory (both additive export-line changes). Confirmed via `git status --porcelain` before commit that `mcp-server.ts`, `session-manager.ts`, `schemas/embedded-agent.ts`, and everything under `packages/client/` were untouched.

## Design decisions not explicitly pinned by the AC (documented here, not silently assumed)

- **Link targets**: `artifact-created` → `/artifacts/{id}` (the real server-side viewer route). `worktree-deletion-finished` → `/worktree-deletion-tasks/{jobId}` (the existing client task-status page, keyed by the job's own id). Spec §11 listed "whether worktree-deletion-finished links to the task page or the repository view" as an open item at AC time — resolved in favor of the task page since it's self-contained (no repo lookup needed) and already exists as a route.
- **`occurredAt` for a stalled job**: job rows have no "became stalled at" timestamp (only `completed_at`, set solely on success). Used `completed_at ?? started_at ?? created_at` as the best available proxy — `started_at` reflects the last processing attempt.
- **Per-source fetch cap (artifacts only)**: `findByUserId(...).slice(0, 50)` — fetch-then-slice rather than pushing a limit into `ArtifactRepository`'s interface (keeps the touched-file surface minimal, per the collision map). Matches R4.6's explicitly accepted v1 cut ("a chatty source can crowd out the other beyond the cap"). **The jobs composer does NOT apply a pre-filter fetch limit** — see the Architect follow-ups section below for why an earlier version of this PR had a real bug here.

## Architect independent audit — 1 required correctness fix, 2 minor hardenings

The Architect ran an independent code-appropriateness audit (in parallel with, and not blocked by, the CodeRabbit wait below) and returned CLEAN-WITH-REQUIRED-FOLLOWUPS. All 3 items are fixed in this PR (see commit `d1265c1c`):

1. **Real bug, multi-user (fixed via TDD, red confirmed before fix):** `composeWorktreeDeletionFinished` fetched the 50 most recent `worktree:delete` jobs **across all users** (the job queue has no per-user scoping) and filtered by `requestUsername` afterward — unlike the artifact composer's `findByUserId`, which is already user-scoped before any slicing. In multi-user mode, once other users collectively produced 50+ more-recent worktree deletions than the caller's own, the caller's own row was evicted from the fetch window before the username filter ever ran — invisible to a single-user dogfood, and distinct from R4.6's accepted "a chatty source can crowd out the other beyond the merged cap" (that cap is applied post-merge in `getFeed`; this was a pre-filter eviction inside one composer). Fix: the jobs composer now fetches by type without a limit (low-frequency job type; the real bound is `getFeed`'s post-merge `FEED_CAP` slice), then filters by status + identity. New regression test: `packages/server/src/services/__tests__/notification-service.test.ts` — `"R-3 fix: caller's own job is not evicted by a fetch-window limit when 51+ OTHER users' newer jobs exist"`; confirmed failing against the pre-fix code (`Expected: true, Received: false`) before applying the fix.
2. **N1 test hardening:** the original N1 deps-shape test only asserted `Object.keys()` on one hand-constructed object, which would not catch a new *optional* member added to `NotificationServiceDeps` (an optional member type-checks and runs fine without ever being supplied). Added a compile-time-only `AssertExactKeys<NotificationServiceDeps, 'artifactRepository' | 'cursorRepository' | 'jobs'>` type assertion in the same test file; confirmed it actually catches the gap by temporarily adding a fake `broadcastToApp?: () => void` to the interface and observing `bun run typecheck` fail (`TS2769`), then reverting.
3. **R3 residual note completeness:** added above.

## Test plan

- [x] `bun run typecheck` — clean across all 5 workspaces.
- [x] `bun run test` — full suite, 0 failures (see below).
- [x] `node .claude/skills/orchestrator/preflight-check.js` — 0 missing sibling tests, 0 rule-duplication, 0 language violations, 0 blame-shift violations.
- [x] Every R4 boundary is a named test:
  1. Tie at cursor (strict `>`) — `packages/server/src/services/__tests__/notification-service.test.ts:153`
  2. Tie between items, deterministic sort — `:162` (same kind), `:177` (different kind)
  3. Mixed-source timestamp normalization (ISO artifact + epoch-ms job) — `:194`
  4. Corrupt job payload: unparseable JSON `:211`, valid-but-non-narrowing shape `:232` — both skip-with-warn, never throw
  5. Non-terminal (`pending`/`processing`) jobs excluded — `:252`
  6. Cap applied after merge+sort — `:263`
  7. `unreadCount` computed pre-cap, exceeds `items.length` at 51+ unread — `:263` (same test)
  8. Invalid `lastSeenAt` format → 400 — `packages/server/src/routes/__tests__/notifications.test.ts:166`; future-timestamp → 400 at `:177`; R2 no-op-on-older at `:187`
- [x] N1 deps-shape test — `packages/server/src/services/__tests__/notification-service.test.ts:114`
- [x] R2 repository-level contract (backward/equal no-op, concurrent-order convergence) — `packages/server/src/repositories/__tests__/sqlite-notification-cursor-repository.test.ts`
- [x] Migration v31 (create + idempotency + round-trip) — `packages/server/src/database/__tests__/migration-v31.test.ts`
- [x] Route auth (401 unauthenticated on both GET and PUT) — `packages/server/src/routes/__tests__/notifications.test.ts`
- [x] Q10 wire-boundary integration test (real route → JSON → schema parse, both GET and PUT) — `packages/integration/src/notification-boundary.test.ts`

### Full test suite output (final run, exit code pasted per workflow.md)

```
shared:         651 pass, 0 fail  (24 files)
server:        4078 pass, 1 skip, 0 fail  (184 files)
embedded-agent: 375 pass, 0 fail
client:        2200 pass, 0 fail
integration:     78+3 pass, 0 fail
scripts/hooks:  538 pass, 0 fail
TEST_EXIT: 0
```

## CodeRabbit

**Note on CodeRabbit (2026-08-18):**

- **Local CLI**: `coderabbit auth login` returned `Non-interactive environment detected. Use --api-key for authentication.` — a headless-worktree auth failure (structural, per `coderabbit-ops` skill's troubleshooting Q2, shape A), not a rate-limit. Not retried, per that disposition.
- **GitHub-side bot**: at the initial head `0ade8953`, commit status read `state=success, description="Review rate limited"` (repo-wide 1-review/hour window consumed by a concurrently open PR). The Orchestrator elected to wait for the window rather than substitute an Architect dual-clean, since the Architect's independent audit (see above) was already in flight in parallel and merge is owner-gated regardless. The `d1265c1c` push (the Architect follow-up fixes) is expected to retrigger a review once the window frees. Disposition owned by the Orchestrator, not this delegate.

## Out of scope / not in this PR

Bell/panel/badge UI (Phase 2, separate Issue); any additional notification source; the obligations lane (§6); a global team feed (§8) — all per the Issue's stated non-goals.

Closes #1353

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added an authenticated notification feed for artifact activity and completed or stalled tasks.
  * Added unread counts across the full notification history, even when only 50 items are displayed.
  * Added support for marking notifications as seen with a persisted read timestamp.
  * Added validation for notification responses and seen-status updates, including timestamp normalization.

* **Documentation**
  * Updated notification API documentation to include unread counts and read-cursor behavior.

---

## CodeRabbit disposition at merge time (recorded by the Orchestrator, 2026-08-18)

**What was reviewed, and when.** This PR received a genuine CodeRabbit review at head `d1265c1c` (17:20 JST) producing 1 actionable Major (the `lastSeenAt` canonicalization) and 3 nitpicks, and a second genuine review at head `11e764e9` (17:31 JST) after the fixes — commit-status description `Review completed`, walkthrough re-posted, **zero new actionable inline comments**, plus one 🟡 Minor that appeared ONLY in the formal review body (auxiliary test databases not closed in `finally`; CodeRabbit noted it could not be posted inline because the code is outside the diff range).

All of it was addressed:
- Major → `321c2672` (route-level `new Date(parsedTime).toISOString()` canonicalization + 2 regression tests, verified failing against the pre-fix code first). CodeRabbit replied on the thread confirming the fix and marked it `✅ Review thread resolved`.
- 3 nitpicks → all adopted in the same commit. One of them was substantive rather than cosmetic: the integration boundary test's `jobs` fake always returned `[]`, so the optional `outcome` field never traversed the parse path — the Q10 guard this file claims to provide was not actually covering its one optional field until this fix.
- Minor (outside-diff) → `58b70829`.

**State now:** the commit status for the current head `58b70829` reads `Review rate limited` — the plan's quota is one review per hour and both of the above consumed it.

**The delta since the last genuine review** (`11e764e9..58b70829`) is **one test file**: `sqlite-notification-cursor-repository.test.ts`, wrapping two auxiliary-database lifecycles in `try/finally` so `destroy()` runs even if an assertion throws. No production code, no schema, no test assertions changed — verified by reading `git diff 11e764e9..58b70829` directly.

**Disposition:** treated as clean-equivalent for merge purposes. The unreviewed delta is exactly the fix CodeRabbit itself asked for, it cannot affect runtime behavior, and every other verdict layer covers the full PR: CI rollup green on the current head, Architect independent audit across two rounds (round 2: `CLEAN-WITH-REQUIRED-FOLLOWUPS`, its single required item landed in `11e764e9`), and the Orchestrator's Q1-Q12 acceptance check PASS. Merge itself remains owner-gated (logic change).

**Operational note for future readers:** two CodeRabbit surfaces behaved in ways that a quick check would misread. (1) The walkthrough is a single comment **edited in place** as the head advances, so an earlier clean verdict leaves no trace in the comment history — it must be recorded when observed. (2) A finding outside the diff range appears **only in the formal review body**, never as an inline comment, so `pulls/N/comments` returning zero does not mean zero findings.

### CODE REVIEW

**Comment by ms2sato** [comment]
@coderabbitai review

File: packages/server/src/routes/notifications.ts [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    // R2: future-timestamp rejection is the route's job -- the SQL WHERE
+    // guard in the repository only compares against the stored value, not
+    // against wall-clock "now".
+    const parsedTime = Date.parse(lastSeenAt);
+    if (Number.isNaN(parsedTime) || parsedTime > Date.now() + FUTURE_TIMESTAMP_TOLERANCE_MS) {
+      throw new ValidationError('lastSeenAt must not be in the future');
+    }
+
+    const current = await notificationCursorRepository.advance(authUser.id, lastSeenAt);
+    return c.json({ lastSeenAt: current });
```
[the remark this answers is not in the document]
Comment (ms2sato) [reply packages/server/src/routes/notifications.ts:79]: Resolved in 321c2672: the route now canonicalizes the validated timestamp with `new Date(parsedTime).toISOString()` before passing it to the cursor repository, so the TEXT comparison is always over UTC-normalized values. Two regression tests were added (canonicalization itself, and a case where lexical and chronological order disagree), both verified failing against the pre-fix code first. The two sibling sites were left unchanged per your note — artifact `createdAt` normalization is unrelated, and the route is the correct fix site.

**Comment by ms2sato** [comment]
@coderabbitai review

**Comment by ms2sato** [comment]
@coderabbitai review
